dataflow/gemma-flex-template/custom_model_gemma.py

```diff
@@ -0,0 +1,178 @@
+from gemma.model import GemmaForCausalLM
+
+import torch
+
+
+class GemmaPytorchModelHandler(ModelHandler[str, PredictionResult,
+                                            GemmaForCausalLM]):
+    def __init__(self,
+                 model_variant: str,
+                 checkpoint_path: str,
+                 tokenizer_path: str,
+                 device: Optional[str] = 'cpu'):
+        """ Implementation of the ModelHandler interface for Gemma-on-Pytorch
+        using text as input.
+
+        Example Usage::
+
+          pcoll | RunInference(GemmaPytorchHandler())
+
+        Args:
+          model_variant: The Gemma model name.
+          checkpoint_path: the path to a local copy of gemma model weights.
+          tokenizer_path: the path to a local copy of the gemma tokenizer
+          device: optional. the device to run inference on. can be either
+            'cpu' or 'gpu', defaults to cpu.
+        """
+        model_config = get_config_for_2b(
+        ) if "2b" in model_variant else get_config_for_7b()
+        model_config.tokenizer = tokenizer_path
+        model_config.quant = 'quant' in model_variant
+        model_config.tokenizer = tokenizer_path
+
+        self._model_config = model_config
+        self._checkpoint_path = checkpoint_path
+        if device == 'GPU':
+            logging.info("Device is set to CUDA")
+            self._device = torch.device('cuda')
+        else:
+            logging.info("Device is set to CPU")
+            self._device = torch.device('cpu')
+        self._env_vars = {}
+
+    def share_model_across_processes(self) -> bool:
+        """Allows us to load a model only once per worker VM, decreasing
+        pipeline memory requirements.
+        """
+        return True
+
+    def load_model(self) -> GemmaForCausalLM:
+        """Loads and initializes a model for processing."""
+        torch.set_default_dtype(self._model_config.get_dtype())
+        model = GemmaForCausalLM(self._model_config)
+        model.load_weights(self._checkpoint_path)
+        model = model.to(self._device).eval()
+        return model
+
+    def run_inference(
+        self,
+        batch: Sequence[str],
+        model: GemmaForCausalLM,
+        inference_args: Optional[Dict[str, Any]] = None
+    ) -> Iterable[PredictionResult]:
+        """Runs inferences on a batch of text strings.
+
+        Args:
+          batch: A sequence of examples as text strings.
+          model: The Gemma model being used.
+          inference_args: Any additional arguments for an inference.
+
+        Returns:
+          An Iterable of type PredictionResult.
+        """
+        # Loop each text string, and use a tuple to store the inference results.
+        predictions = []
+        result = model.generate(prompts=batch, device=self._device)
+        predictions.append(result)
+        return [PredictionResult(x, y) for x, y in zip(batch, predictions)]
+
+
+class FormatOutput(beam.DoFn):
```

Comment:
This could be a simple function rather than an entire class.

```py
def format_output(element):
  return "Input: {input}, Output: {output}".format(
      input=element.example, output=element.inference)
```

And then use `beam.Map` instead of `beam.ParDo`:

```py
    | "Format Output" >> beam.Map(format_output)
```